docs/user_utilities.rst

```diff
@@ -0,0 +1,109 @@
+User utilities
+==============
+
+Taichi provides a number of development tools that can be used to greatly improve development efficiency. This section aims to provide a detailed description of some of the commonly used sections. Including:
+
+* Logging
+* Profiler
+* Assertation
+
+Logging
+-------
+
+Based on spdlog, Taichi provides a series of APIs for printing logs; these functions can be used for recording during Taichi's runtime.
+
+.. Note::
+
+    The current taichi logging API only supports standard output, not output to files.
+
+.. function:: ti.set_logging_level(level)
+
+    :parameter level: (string) a valid logging level
+
+    - This function is used to set the level of print logs. Currently the log levels in taichi are classified from low to high as ``ti.TRACE``, ``ti.DEBUG``, ``ti.INFO``, ``ti.WARN``, ``ti.ERR`` and ``ti.CRITICAL``. These logging levels are essentially strings "trace", "debug", "info", "warn", "error" and "critical" respectively.
+    
+    - The lower the log level, the more content will be printed out. For example, if you set the print level to ``ti.TRACE``, all logs will be printed; if you set the print level to ``ti.ERR``, only those log which is generated by `ti.critical()` and ``ti.error()`` will be output.
+    
+    - The default logging level is ``ti.INFO``.You can also override the default logging level by setting the environment variable like ``TI_LOG_LEVEL=warn``.
+
+.. function:: ti.info(info)
+
+    :parameter info: (string) logging info
+
+    Print out the input string info to stdout if the logging level is lower or equal to ``ti.INFO``. For example:
+
+.. code-block:: python
+
+    import taichi as ti
+    ti.init(arch = ti.cpu)
+
+    ti.set_logging_level(ti.INFO)
+    var = ti.var(ti.f32, shape = 1)
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+        ti.info(f"set var[0] = 1.0")
```

Comment:
Let's be clear that this is only called at compile-time. i.e. If you recall compute, it won't reprint the string. (same as ti.static_print)

Reply:
Agree. 

I have an observation, would you please help me to check whether my understanding is correct?

According to the `Debugging` part in this [doc](https://taichi.readthedocs.io/en/stable/syntax.html), It seems that all of the `print` will be transfered to `ti.static_print` by the AST preprocessor, so that they will be executed in run-time. That's good.

But it doesn't include logging utils such `ti.info`, these `ti.info` are only called at compiled-time but not runtime.
***
So we cannot do any logging `ti.info` in kernel runtime, right? :( only print is supported in runtime is a little trivial, and IMO the debugging experince can be nicer if we support `ti.info` in runtime.

Basically, why wasn't `ti.info` be supported in runtime? It looks not very hard. Would you be happy to introduce the idea behind this design? :D

Reply:
> It looks not very hard

No, **copying strings from GPU to CPU** is harder than you may think. Ancient GPUs doesn't even support integer, not to mention to implement runtime f-string...

In fact, the only difference between `ti.info` and `print` is its color and f-string.. not sure why `ti.info` superior.

Reply:
Thanks for your clear explanation! I get the point. Yes, I fully understand your meaning. ;-) 

The difference is just the timestamp(color is relatively unimportant). Let's assume we want to do a simulation which can cost for a whole day (It's quite common in mechanical and chemstry simulation, most of them are in reseach cases), If we want to find the bug after waking up in the morning, a timestamp or screenshot can be helpful IMAO.

Reply:
I think they could do `ti.info` **between kernel launches**.
In fact, `print` in kernel also won't print until kernel ends as my explanation goes above.
So there's no difference for this usage.

If you want a timestamp in `print`, I think `print(ti.timestamp(), 'x =', x)` is a good approach (TBD).

Reply:
Thanks! 

I will invesigate this case further more. I need to get a global view of Taichi in order to understand it, instead of floating on the surface. XD